crypto/mbedtls/source/bignum_alt.c

```diff
@@ -0,0 +1,299 @@
+
+  fd = open("/dev/mpi0", O_RDWR);
+  if (fd < 0)
+    {
+      return -errno;
+    }
+
+  mpi.op = MPI_CALC_FUNC_SUB;
+  mbedtls_mpi_to_mpiparam(&mpi.param[0], A);
+  mbedtls_mpi_to_mpiparam(&mpi.param[1], B);
+
+  mbedtls_mpi_grow(X, MBEDTLS_ROUNDUP(MAX(A->n, B->n) + 1,
+                                              sizeof(mbedtls_mpi_uint)));
+  mbedtls_mpi_to_mpiparam(&mpi.param[2], X);
+  ret = ioctl(fd, MATHIOC_MPI_CALC, (unsigned long)((uintptr_t)&mpi));
+  if (ret >= 0)
+    {
+      mpiparam_to_mbedtls_mpi(X, &mpi.param[2]);
+    }
+
+  close(fd);
+  return ret;
+}
+
+int mbedtls_mpi_mul_mpi(FAR mbedtls_mpi *X, FAR const mbedtls_mpi *A,
+                        FAR const mbedtls_mpi *B)
+{
+  int ret = 0;
+  int fd;
+  struct mpi_calc_s mpi;
+
+  fd = open("/dev/mpi0", O_RDWR);
+  if (fd < 0)
+    {
+      return -errno;
+    }
+
+  mpi.op = MPI_CALC_FUNC_MUL;
+  mbedtls_mpi_to_mpiparam(&mpi.param[0], A);
+  mbedtls_mpi_to_mpiparam(&mpi.param[1], B);
+
+  mbedtls_mpi_grow(X, MBEDTLS_ROUNDUP(A->n + B->n,
+                                              sizeof(mbedtls_mpi_uint)));
+  mbedtls_mpi_to_mpiparam(&mpi.param[2], X);
+  ret = ioctl(fd, MATHIOC_MPI_CALC, (unsigned long)((uintptr_t)&mpi));
+  if (ret >= 0)
+    {
+      mpiparam_to_mbedtls_mpi(X, &mpi.param[2]);
+    }
+
+  close(fd);
+  return ret;
+}
+
+int mbedtls_mpi_div_mpi(FAR mbedtls_mpi *Q, FAR mbedtls_mpi *R,
+                        FAR const mbedtls_mpi *A, FAR const mbedtls_mpi *B)
+{
+  int ret = 0;
+  int fd;
+  struct mpi_calc_s mpi;
+
+  fd = open("/dev/mpi0", O_RDWR);
+  if (fd < 0)
+    {
+      return -errno;
+    }
+
+  mpi.op = MPI_CALC_FUNC_DIV;
+  mbedtls_mpi_to_mpiparam(&mpi.param[0], A);
+  mbedtls_mpi_to_mpiparam(&mpi.param[1], B);
+  mbedtls_mpi_grow(Q, A->n);
+  mbedtls_mpi_grow(R, B->n);
+  mbedtls_mpi_to_mpiparam(&mpi.param[2], Q);
+  mbedtls_mpi_to_mpiparam(&mpi.param[3], R);
+  ret = ioctl(fd, MATHIOC_MPI_CALC, (unsigned long)((uintptr_t)&mpi));
+  if (ret >= 0)
+    {
+      mpiparam_to_mbedtls_mpi(Q, &mpi.param[2]);
+      mpiparam_to_mbedtls_mpi(R, &mpi.param[3]);
+    }
+
+  close(fd);
+  return ret;
+}
+
+int mbedtls_mpi_mod_mpi(FAR mbedtls_mpi *R, FAR const mbedtls_mpi *A,
+                        FAR const mbedtls_mpi *B)
+{
+  int ret = 0;
+  int fd;
+  struct mpi_calc_s mpi;
+
+  fd = open("/dev/mpi0", O_RDWR);
+  if (fd < 0)
+    {
+      return -errno;
+    }
+
+  mpi.op = MPI_CALC_FUNC_MOD;
+  mbedtls_mpi_to_mpiparam(&mpi.param[0], A);
+  mbedtls_mpi_to_mpiparam(&mpi.param[1], B);
+  mbedtls_mpi_grow(R, B->n);
+  mbedtls_mpi_to_mpiparam(&mpi.param[2], R);
+  ret = ioctl(fd, MATHIOC_MPI_CALC, (unsigned long)((uintptr_t)&mpi));
+  if (ret >= 0)
+    {
+      mpiparam_to_mbedtls_mpi(R, &mpi.param[2]);
+    }
+
+  close(fd);
+  return ret;
+}
+
+int mbedtls_mpi_exp_mod(FAR mbedtls_mpi *X, FAR const mbedtls_mpi *A,
+                        FAR const mbedtls_mpi *E, FAR const mbedtls_mpi *N,
+                        FAR mbedtls_mpi *)
```

Comment:
What is the name of the last parameter here? Is it used?

Reply:
the last parameter of origin is prec_RR, which is a calculation result that can reuse through the previous exponentiation operation. This cannot be used without the mbedtls buildin software algorithm.